### PR TITLE
schemafeed: Add an optimization case where we don't wait on Peek or Pop

### PR DESCRIPTION
**Pull request by Xiang-Gu**
This commit utilizes the recently introduced storage parameter
    "schema_locked" to achieve low-latency changefeed. Namely, we added logic to Peek/Pop to return "there is no table events" if the table schemas are locked without waiting for `highWater` to reach `atOrBefore`. 
    This change allows us to reduce event commit-to-emit latency drastically
    from `changefeed.experimental_poll_interval` (which defaults to 1s) to
    milliseconds.

Epic: CRDB-2258
Release note: None

### CODE REVIEW

**Review by miretskiy — COMMENTED** [review]
we definitely want to have rel notes for changefeed -- low latency stuff is a big deal.
Epic CRDB-2258

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-:-NTL0y9h4ewJZj80i6Kp:b-riagul)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Xiang-Gu)

___
*[`pkg/ccl/changefeedccl/kvfeed/kv_feed.go` line 360 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTL18u41geHAiT_8P2p:-NTL18u41geHAiT_8P2q:b-xbj4f9) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L360)):*
> ```Go
> }
> 
> // scanIfShould performs a scan of KV paris in watched span if
> ```

s/paris/pairs/
___
*[`pkg/ccl/changefeedccl/kvfeed/kv_feed.go` line 504 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTL1IPH9gztAWowOcy7:-NTL1IPH9gztAWowOcy8:b-v3ct8i) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L504)):*
> ```Go
> 	// - `copyFromSourceToDestUntilTableEvent` consumes `membuf` into `f.writer`
> 	// until a table event (i.e. a column is added/dropped) has occurred, which
> 	// signals another possible scan.
> ```

Thanks for making existing code better by adding comments.
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTL0YJb4PHU-M8tajXl:-NTL0YJb4PHU-M8tajXm:b-c1e8p0) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L311)):*
> ```Go
> func (tf *schemaFeed) pollTableHistory(ctx context.Context) error {
> 	for {
> 		if !tf.pollingPaused() {
> ```

This is just great.
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 312 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTL1Xhb1QzsrW84dV9X:-NTL1Xhb1QzsrW84dV9Y:bz1739l) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L312)):*
> ```Go
> 	for {
> 		if !tf.pollingPaused() {
> 			if err := tf.updateTableHistory(ctx, tf.clock.Now()); err != nil {
> ```

mind extending metrics to add a counter on the number of "expensive" scans?
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 404 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTKcuXb1qxc-GJz0tcs:-NTKcuXb1qxc-GJz0tct:bn8r51i) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L404)):*
> ```Go
> // This means we can use this predecessor version as if it's the canonical
> // version in the changefeed setup, since those two versions are "identical"
> // as far as triggering changefeed scans or decoding KV events are concerned.
> ```

:heart: this comment. super nice.
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 417 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTKdFH89JKAe6E429Vk:-NTKdFH97LQ8ujYuk8rG:b-m67q5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L417)):*
> ```Go
> 			}
> 			if !ld.Underlying().(catalog.TableDescriptor).IsSchemaLocked() {
> 				allWatchedTableSchemaLocked = false
> ```

should we return an error here to break iteration early if we find non-locked descriptor?
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 440 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTKykgrBFEsMQ5wJTrD:-NTKykgrBFEsMQ5wJTrE:b-m0erdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L440)):*
> ```Go
> 	}
> 	if !ok1 || !ok2 || len(versions1) != len(versions2) {
> 		tf.mu.pollingPaused = false
> ```

should this be set unconditionally after we acquire the lock?
and set to true only if we know that it can be paused (below)?

**Review by Xiang-Gu — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101694)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[`pkg/ccl/changefeedccl/kvfeed/kv_feed.go` line 360 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTL18u41geHAiT_8P2p:-NTP-xdm7V9vTYZAa4Ri:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L360)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

s/paris/pairs/
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/kvfeed/kv_feed.go` line 504 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTL1IPH9gztAWowOcy7:-NTP-yGcDRQi8ZqCd5b2:b32vhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L504)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Thanks for making existing code better by adding comments.
</blockquote></details>

:)
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTL0YJb4PHU-M8tajXl:-NTP-zJ4Byvduv6j7o3H:bmi8ppe) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L311)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This is just great.
</blockquote></details>

:100: 
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 312 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTL1Xhb1QzsrW84dV9X:-NTQFkasBTn8_Nmnol5h:ba23g82) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L312)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

mind extending metrics to add a counter on the number of "expensive" scans?
</blockquote></details>

Ack, going to do it soon
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 404 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTKcuXb1qxc-GJz0tcs:-NTP04A52T86WnVXg1gQ:b32vhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L404)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

:heart: this comment. super nice.
</blockquote></details>

:)
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 417 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTKdFH89JKAe6E429Vk:-NTP2XAV6AKi-g47CRCQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L417)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

should we return an error here to break iteration early if we find non-locked descriptor?
</blockquote></details>

Done.



_Code quote:_
```Go
!ld.Underlying().(catalog.TableDescriptor).
```
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 440 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTKykgrBFEsMQ5wJTrD:-NTP4EAm7qEhV_mo48fT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L440)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

should this be set unconditionally after we acquire the lock?
and set to true only if we know that it can be paused (below)?
</blockquote></details>

Done.



_Code quote:_
```Go
tf.mu.pollingPaused = false
```

**Comment by Xiang-Gu** [comment]
I tried to run the kv workload and see the commit latency. Here is what I had for running the workload for 30s with the table unlocked and locked:
Screenshot 2023-04-19 at 4 39 07 PM


Screenshot 2023-04-19 at 4 43 24 PM


Looks like it's working 😄 

I'm going to add a metric to count the number of "expansive scan" performed by the schemaFeed polling next.

**Review by Xiang-Gu — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101694)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 312 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTL1Xhb1QzsrW84dV9X:-NTV-TdkBEMFu8V1M7kO:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b35ac1ac82c7ba93520c54a9e5394c8c7810e2/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L312)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

Ack, going to do it soon
</blockquote></details>

done

**Review by miretskiy — APPROVED** [review]
> Reviewed 1 of 10 files at r5, 3 of 7 files at r6, 1 of 5 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101694)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Xiang-Gu)

___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 444 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTV-uP44QMEtwyeYkjm:-NTV-uP44QMEtwyeYkjn:b-rpn516) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9f0f19773a619ea8a0d5ff5192fc3a01f25388/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L444)):*
> ```Go
> 			// If a table is dropped and cause Acquire to fail, we mark it as terminal
> 			// error, so we don't retry and let the changefeed job handle this error.
> 			err = changefeedbase.WithTerminalError(err)
> ```

perfect.

**Review by miretskiy — COMMENTED** [review]
One more request: Can you significantly expand the commit message, and add release notes?
We want to show commands to run to lock the schema/ to unlock when needed, include release note
so that we get this into docs...

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-:-NTV3ZtA6pZYoRKhbdqq:b-aejg46)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Xiang-Gu)

**Comment by Xiang-Gu** [comment]
@miretskiy I recently made some changes to this PR and I want you to take a look another look at it. In the first glance, it looks "more"/"harder-to-understand" but I am happier with this new version because it respects the peekOrPop contract more and makes less assumptions.

**Review by miretskiy — CHANGES_REQUESTED** [review]
Can you rebase since the schema portion already merged?

> Reviewed 3 of 7 files at r6, 2 of 49 files at r8, 4 of 49 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-:-NUwqVkZCvQKUkSN70Rj:b-bu0yia)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @shermanCRL and @Xiang-Gu)

___
*[`pkg/ccl/changefeedccl/schemafeed/metrics.go` line 34 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NUwZv-Y3af986JDMyZa:-NUwZv-Y3af986JDMyZb:b3gheca) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d719d731075144200650ed11f632847596ba042/pkg/ccl/changefeedccl/schemafeed/metrics.go#L34)):*
> ```Go
> type Metrics struct {
> 	TableMetadataNanos *metric.Counter
> 	TableHistoryScans  *metric.Counter
> ```

should we add a "TableHistoryCacheHits" or something like that to reflect that fast path?
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 442 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NUwp4KACbBFmc82S7SK:-NUwp4KACbBFmc82S7SL:book8ea) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d719d731075144200650ed11f632847596ba042/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L442)):*
> ```Go
> 	areAllLeasedTablesSchemaLocked := func(ts hlc.Timestamp) (bool, map[descpb.ID]descpb.DescriptorVersion, error) {
> 		allWatchedTableSchemaLocked := true
> 		allWatchedTableVersion := make(map[descpb.ID]descpb.DescriptorVersion)
> ```

This code will run (very) frequently -- for each event.
Even though this map will almost always be a single element map, even single
element map allocation would be costly.
we should memoize this map in the schemaFeed struct.
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 480 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NUwpt8n08BEs2PqP94h:-NUwpt8n08BEs2PqP94i:b-v2n546) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d719d731075144200650ed11f632847596ba042/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L480)):*
> ```Go
> 		return atOrBefore, err
> 	}
> 	if !ok1 || !ok2 || !reflect.DeepEqual(versions1, versions2) {
> ```

would be nice not to use reflect on a hot path.

wouldn't comparison of descriptor versions be sufficient?

**Review by Xiang-Gu — COMMENTED** [review]
I've addressed all your comments and I think it's ready for another look!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-:-NV0liNu-fmt279Fe-uG:b7b1fsf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @shermanCRL)

___
*[`pkg/ccl/changefeedccl/schemafeed/metrics.go` line 34 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NUwZv-Y3af986JDMyZa:-NV0VxfT0zFm00kbFDtU:b-lcpdau) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d719d731075144200650ed11f632847596ba042/pkg/ccl/changefeedccl/schemafeed/metrics.go#L34)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

should we add a "TableHistoryCacheHits" or something like that to reflect that fast path?
</blockquote></details>

Doesn't this newly added metric `TableHistoryScans` reflect that?
Every time an event is committed, we exercise the check to either pause or resume the history scan (or, put differently, either we hit the fast path or not), and I think `TableHistoryScans` is a good metric for us to know that.
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 444 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NTV-uP44QMEtwyeYkjm:-NTV_cCU35P0ibInNWpO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9f0f19773a619ea8a0d5ff5192fc3a01f25388/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L444)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

perfect.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 442 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NUwp4KACbBFmc82S7SK:-NV0gh891MSwc0rIf_94:bsmtpvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d719d731075144200650ed11f632847596ba042/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L442)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This code will run (very) frequently -- for each event.
Even though this map will almost always be a single element map, even single
element map allocation would be costly.
we should memoize this map in the schemaFeed struct.
</blockquote></details>

Good point! I've added two maps in the schemaFeed struct initialized once duration initialization and is used here to avoid allocating a map every time this hot path is called.
___
*[`pkg/ccl/changefeedccl/schemafeed/schema_feed.go` line 480 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NUwpt8n08BEs2PqP94h:-NV0gUfHAQmDFe5V6JEd:b5dp8q8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d719d731075144200650ed11f632847596ba042/pkg/ccl/changefeedccl/schemafeed/schema_feed.go#L480)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

would be nice not to use reflect on a hot path.

wouldn't comparison of descriptor versions be sufficient?

</blockquote></details>

done. I used reflect simply to get multiple lines into one line but I see the performance implication now so I reverted back to a simple for-loop comparision.

**Review by miretskiy — APPROVED** [review]
> Reviewed 1 of 4 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101694)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @shermanCRL and @Xiang-Gu)

___
*[`pkg/ccl/changefeedccl/schemafeed/metrics.go` line 34 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/101694#-NUwZv-Y3af986JDMyZa:-NV0m24XEZixWCoCYQhO:bsntft0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d719d731075144200650ed11f632847596ba042/pkg/ccl/changefeedccl/schemafeed/metrics.go#L34)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

Doesn't this newly added metric `TableHistoryScans` reflect that?
Every time an event is committed, we exercise the check to either pause or resume the history scan (or, put differently, either we hit the fast path or not), and I think `TableHistoryScans` is a good metric for us to know that.
</blockquote></details>

I thought that scans is the expensive part, and suggested cache hits to keep track of fast path.
Basically, I want to see how often we do expensive scan (hopefully never).
It's okay if you feel this metric is sufficient.
